File: commerce_app/lib/models/coupon.dart
class Coupon {
  final int id;
  final int userId;
  final int offerId;
  final String couponCode;
  final String? qrCodeData;
  final String status;
  final String? generatedAt;
  final String? expiresAt;
  final String? usedAt;

  Coupon({
    required this.id,
    required this.userId,
    required this.offerId,
    required this.couponCode,
    this.qrCodeData,
    required this.status,
    this.generatedAt,
    this.expiresAt,
    this.usedAt,
  });

  factory Coupon.fromJson(Map<String, dynamic> json) {
    return Coupon(
      id: (json['id'] as num).toInt(),
      userId: (json['userId'] as num).toInt(),
      offerId: (json['offerId'] as num).toInt(),
      couponCode: json['couponCode'] as String,
      qrCodeData: json['qrCodeData'] as String?,
      status: json['status'] as String? ?? 'GENERATED',
      generatedAt: json['generatedAt'] as String?,
      expiresAt: json['expiresAt'] as String?,
      usedAt: json['usedAt'] as String?,
    );
  }

  bool get isValid => status == 'USED'; // après validation côté commerce
}

File: backend/src/main/java/com/campuspass/backend/model/Coupon.java
package com.campuspass.backend.model;

import com.campuspass.backend.model.enums.CouponStatus;
import com.campuspass.backend.model.User;
import jakarta.persistence.*;
import java.time.LocalDateTime;

@Entity
@Table(name = "coupons")
public class Coupon {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private Long userId;
    @Column(nullable = false)
    private Long offerId;

    @Column(nullable = false, unique = true)
    private String couponCode;
    private String qrCodeData;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CouponStatus status = CouponStatus.GENERATED;

    private LocalDateTime generatedAt;
    private LocalDateTime expiresAt;
    private LocalDateTime usedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "userId", insertable = false, updatable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "offerId", insertable = false, updatable = false)
    private Offer offer;

    public Coupon() {}

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public Long getUserId() { return userId; }
    public void setUserId(Long userId) { this.userId = userId; }
    public Long getOfferId() { return offerId; }
    public void setOfferId(Long offerId) { this.offerId = offerId; }
    public String getCouponCode() { return couponCode; }
    public void setCouponCode(String couponCode) { this.couponCode = couponCode; }
    public String getQrCodeData() { return qrCodeData; }
    public void setQrCodeData(String qrCodeData) { this.qrCodeData = qrCodeData; }
    public CouponStatus getStatus() { return status; }
    public void setStatus(CouponStatus status) { this.status = status; }
    public LocalDateTime getGeneratedAt() { return generatedAt; }
    public void setGeneratedAt(LocalDateTime generatedAt) { this.generatedAt = generatedAt; }
    public LocalDateTime getExpiresAt() { return expiresAt; }
    public void setExpiresAt(LocalDateTime expiresAt) { this.expiresAt = expiresAt; }
    public LocalDateTime getUsedAt() { return usedAt; }
    public void setUsedAt(LocalDateTime usedAt) { this.usedAt = usedAt; }
}
